components/appmgt/org.wso2.carbon.appmgt.custom.mediators/src/main/java/org/wso2/carbon/appmgt/custom/mediators/redirection/RefererRewriteMediator.java

```diff
@@ -40,32 +37,31 @@
      */
     public boolean mediate(MessageContext messageContext) {
 
-        Cache cache = this.getAppContextVersionConfigCache();
-        Map<String,String> contextVersionMap = (HashMap<String, String>) cache.get
-                (AppMConstants.APP_CONTEXT_VERSION_CACHE_KEY);
-        //ToDo: handle if cache has expired
-
-        org.apache.axis2.context.MessageContext axis2MC =
-                ((Axis2MessageContext) messageContext). getAxis2MessageContext();
-        Map map = (Map)axis2MC.getProperty(AppMConstants.TRANSPORT_HEADERS);
-                String host = (String) map.get(AppMConstants.HOST);
-        String referer = (String)map.get(AppMConstants.REFERER);
+        Map<String, String> contextVersionMap = AppContextCacheUtil.getTenantContextVersionUrlMap();
+        org.apache.axis2.context.MessageContext axis2MC = ((Axis2MessageContext) messageContext)
+                .getAxis2MessageContext();
+        Map map = (Map) axis2MC.getProperty(AppMConstants.TRANSPORT_HEADERS);
+        String host = (String) map.get(AppMConstants.HOST);
+        String referer = (String) map.get(AppMConstants.REFERER);
 
         //Identify http/https
         String transport = referer.split(AppMConstants.URL_DELIMITER)[0];
         String endpoint;
         String restFullRequestPath = (String) messageContext.getProperty("REST_FULL_REQUEST_PATH");
-        for (String contextVersion : contextVersionMap.keySet()){
-            if (referer.contains(contextVersion)){
+        for (String contextVersion : contextVersionMap.keySet()) {
+            //TODO: This iteration is too slow. we need a better data structure for this.
```

Comment:
Yes. However I think when detecting a possible performance problem, and there is a time constraint, and not related to original issue being fixed, It is better at least to put a comment saying possible problem, rather than ignore it. 
I could have opened a JIRA, but then, it will become disconnected if this code ever changed.